### PR TITLE
CI: move sdist/pre-release test Azure

### PR DESCRIPTION
**Pull request by tylerjereddy**
* move the sdist/NumPy pre-release wheel
CI entry from Travis CI to Azure CI

Not working yet though, unfortunately.

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -24,6 +24,52 @@ variables:
...
+  steps:
+  - task: UsePythonVersion@0
+    inputs:
+      versionSpec: '3.7'
+      addToPath: true
+      architecture: 'x64'
+  - script: >-
+      python -m pip install --upgrade
+      pip==20.2.4
+      "setuptools<50"
```
Comment (tylerjereddy) [inline azure-pipelines.yml:40]: this shouldn't be needed, and didn't help with the issue anyway..

File: azure-pipelines.yml [hunk 2]
```diff
@@ -24,6 +24,52 @@ variables:
...
+    displayName: 'Install pre-release deps'
+    failOnStderr: true
+  - script: |
+      python -m pip install --pre --upgrade --timeout=60 -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy
+    displayName: 'Install pre-release NumPy wheel'
+    failOnStderr: true
+  - script: |
+        python tools/suppress_output.py python setup.py sdist
+        cd dist
+        python -m pip install $PWD/scipy* -v
```
Comment (tylerjereddy) [inline azure-pipelines.yml:60]: Maybe the build isolation is causing issues for the build? I think the Travis CI comments suggest build isolation is desirable for this CI entry.

Comment (tylerjereddy) [reply azure-pipelines.yml:60]: I'll try to debug on my fork, it may be the sdist setup proper

**Comment by andyfaff** [comment]
It's an ideal time to rationalise the matrix that was on travis. What do the sdist and pre-release tests do? Is that numpy built from source?

**Comment by tylerjereddy** [comment]
> Is that numpy built from source?

Should be one of their "nightly" wheels

**Comment by tylerjereddy** [comment]
Looks like the sdist build failure was because of this fact Pauli pointed out years ago: https://github.com/numpy/numpy/issues/7574#issuecomment-215084166

I've pushed in a revision to checkout the submodules needed for sdist building; apparently, the submodules were available automatically in Travis but not Azure then.

**Comment by tylerjereddy** [comment]
CI is all green and seems to do what is needed from the original Travis run for that matrix entry.

**Comment by rgommers** [comment]
> It's an ideal time to rationalise the matrix that was on travis. What do the sdist and pre-release tests do? Is that numpy built from source?

The sdist one is important, it's the thing that catches packaging-related issues like mistakes in `MANIFEST.in`. Previously those would only be caught once a release candidate hit PyPI and couldn't build because of missing files.

The sdist test can of course be combined with some other property, like testing a specific Python/NumPy version.

The pre-release ones test NumPy master indeed.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -24,6 +24,54 @@ variables:
...
+    vmImage: 'ubuntu-18.04'
+  steps:
+  - task: UsePythonVersion@0
+    inputs:
+      versionSpec: '3.7'
+      addToPath: true
+      architecture: 'x64'
+  - script: >-
+      python -m pip install --upgrade
+      pip==20.2.4
```
Comment (rgommers) [inline azure-pipelines.yml:39]: What's going on here? `pip` is downgraded to 20.2.4 after being upgraded on the line above it looks like, is that necessary?

Comment (rgommers) [reply azure-pipelines.yml:39]: Or wait, is this all one command? If so, why not put it on a single line?

**Review by rgommers — COMMENTED** [review]
This works, but it's very verbose compared to the old version. I don't know all that much about Azure, so if it's necessary then so be it. But won't this get hard to maintain quickly?

File: azure-pipelines.yml (hunk 3 above)
Comment (tylerjereddy) [reply azure-pipelines.yml:39]: The pip version pin is present in the Travis yml as far as I can tell; separate alphabetically-sorted lines for deps is just for readability, similar to apt-get package specifiers in Dockerfiles; we can condense to a single long line--I don't mind..

**Comment by tylerjereddy** [comment]
> But won't this get hard to maintain quickly?

There are two dimensions of verbosity I think--the single line pip commands are just to separate out the deps in alphabetical order for maintainability. I don't care if we condense those to single 200 characters lines or whatever.

The bigger problem is that I have not been using matrices for the new entries I migrate over--I figured we'd migrate over and then refactor jobs that have enough stuff in common, mostly just because I'm working fast..

**Comment by peterbell10** [comment]
It looks like [azure templates](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates) would be the most similar to how the TravisCI script is organized currently: with a single common script, controlled by defining parameters. I could try out a a more direct translation of the travis matrix if that sounds reasonable.

**Comment by tylerjereddy** [comment]
@peterbell10 Ok, I mentioned/pinged you in the migration tracking issue--I can guess folks can coordinate there. If you want to supersede this feel free--I was just panicking because there was a bunch of migration work to do and I wanted to move it forward.

**Comment by rgommers** [comment]
> The bigger problem is that I have not been using matrices for the new entries I migrate over--I figured we'd migrate over and then refactor jobs that have enough stuff in common, mostly just because I'm working fast..

That's what I meant.

We can still merge this as is as long as it's green, and then try to improve on it with an Azure template maybe?

**Comment by tylerjereddy** [comment]
Alright merged, sounds like folks are willing to help refactor

**Pull request merged**